orchagent/routeorch.cpp

```diff
@@ -2547,3 +2565,968 @@ void RouteOrch::decNhgRefCount(const std::string &nhg_index)
+
+        for (auto &entry : iter->second)
+        {
+            createSingleRouteFlowCounterByPattern(route_pattern, entry.first, to_bind, current_bound_count);
+        }
+    }
+
+    auto other_iter = mOtherRouteEntries.find(route_pattern.vrf_id);
+    if (other_iter != mOtherRouteEntries.end())
+    {
+        SWSS_LOG_NOTICE("Creating route flow counter for pattern %s for other type route entries", route_pattern.to_string().c_str());
+
+        for (auto ip_prefix : other_iter->second)
+        {
+            createSingleRouteFlowCounterByPattern(route_pattern, ip_prefix, to_bind, current_bound_count);
+        }
+    }
+
+    flushBindFlowCounters(route_pattern, to_bind, current_bound_count);
+}
+
+void RouteOrch::createSingleRouteFlowCounterByPattern(
+    const RoutePattern &route_pattern,
+    const IpPrefix &ip_prefix,
+    std::list<RouteFlowCounterBulkContext> &to_bind,
+    size_t &current_bound_count)
+{
+    if (route_pattern.is_match(route_pattern.vrf_id, ip_prefix))
+    {
+        if (current_bound_count >= route_pattern.max_match_count)
+        {
+            SWSS_LOG_DEBUG("Adding other type route entry to unbound cache, current bound=%lu, max allowed=%lu", current_bound_count, route_pattern.max_match_count);
+            addUnboundRoutesToCache(route_pattern, ip_prefix);
+            return;
+        }
+
+        if (!bulkBindFlowCounters(route_pattern.vrf_id, ip_prefix, to_bind, current_bound_count))
+        {
+            SWSS_LOG_WARN("Failed to bind route flow counter, vrf=%s, ip_prefix=%s", sai_serialize_object_id(route_pattern.vrf_id).c_str(), ip_prefix.to_string().c_str());
+            return;
+        }
+
+        if (current_bound_count >= route_pattern.max_match_count)
+        {
+            flushBindFlowCounters(route_pattern, to_bind, current_bound_count);
+        }
+    }
+}
+
+void RouteOrch::createRouteFlowCounterFromUnboundCacheByPattern(const RoutePattern &route_pattern, size_t current_bound_count)
+{
+    SWSS_LOG_ENTER();
+
+    auto iter = mUnboundRoutes.find(route_pattern);
+    if (iter == mUnboundRoutes.end())
+    {
+        return;
+    }
+
+    std::list<RouteFlowCounterBulkContext> to_bind;
+
+    while(current_bound_count < route_pattern.max_match_count)
+    {
+        auto prefix_iter = iter->second.begin();
+        if (prefix_iter == iter->second.end())
+        {
+            break;
+        }
+
+        if (!bulkBindFlowCounters(route_pattern.vrf_id, *prefix_iter, to_bind, current_bound_count))
+        {
+            SWSS_LOG_WARN("Failed to bind route flow counter, vrf=%s, ip_prefix=%s", sai_serialize_object_id(route_pattern.vrf_id).c_str(), prefix_iter->to_string().c_str());
+            iter->second.erase(prefix_iter);
+            continue;
+        }
+
+        if (current_bound_count >= route_pattern.max_match_count)
+        {
+            flushBindFlowCounters(route_pattern, to_bind, current_bound_count);
```

Comment:
yes, it is needed because there could be failures when binding counter to a route entry. Once there is a failure, we do "--current_bound_count" inside flushBindFlowCounters. And that means we still have "room" for more route flow counters.

Reply:
What is the purpose of calling it at the end of the function? Is it redundant?

Reply:
In case it did not reach max_match_count in the loop.